src/components/fhir/FhirProvider.tsx

```diff
@@ -0,0 +1,42 @@
+import {oauth2 as SMART} from "fhirclient";
+import {FC, useState} from "react";
+import {FhirContext} from "./FhirContext";
+import Client from "fhirclient/lib/Client";
+
+/**
+ * Context provider to allow FHIR client to be passed through to child components.
+ * @param children children components to render
+ * @constructor
+ */
+const FhirProvider: FC<any> = ({children}) => {
+  const [client, setClient] = useState<Client | null>(null);
+  const [error, setError] = useState<Error | null>(null);
+
+  const setClientOrError = (client: Client | null) => {
+    if (!client) {
+      SMART.ready()
+        .then((client) => setClient(client))
+        .catch((error) => setError(error))
+        .finally(() => console.debug("FHIR client ready"))
+    }
+  };
+
+  if (error) {
+    console.error(error.stack);
+    return <pre>{error.message}</pre>;
+  }
+
+  return (
+    <FhirContext.Provider value={{client: client, setClient: setClient}}>
+      <FhirContext.Consumer>
+        {({client}) => {
+          setClientOrError(client)
+          return children;
+        }}
+      </FhirContext.Consumer>
+    </FhirContext.Provider>
```

Comment:
This seems slightly strange to me (although it may be because I lack knowledge of the way the fhirclient works!). I can see we are using both `FhirContext.Consumer` in addition to the `useContext` hook in your `ReportForm` component. Typically, you would use one or the other. It looks like the consumer here is passing the context to the entire app via `return children` rather than exposing it only to necessary components that need it, is that correct? Then you are also accessing this context in `ReportForm` using hooks. 

Using both methods might be overkill, unless there is some restriction imposed by the fhirclient library?

Reply:
That's a good point, let me have a look at this with a fresh pair of eyes and have a think